File: library/runner-cli/sensemaker_model_cli.ts
import { Command } from "commander";
import { Model } from "../src/models/model";
import { VertexModel } from "../src/models/vertex_model";
import { OllamaModel } from "../src/models/ollama_model";
import { OpenAiCompatModel } from "../src/models/openai_compat_model";

export type SensemakerAdapter = "vertex" | "ollama" | "openai-compatible";
export type OpenAiCompatProvider = "openai" | "openrouter" | "mistral";

export const DEFAULT_VERTEX_MODEL = "gemini-2.5-pro";
export const DEFAULT_VERTEX_LOCATION = "global";
export const DEFAULT_OLLAMA_MODEL = "gemma3:latest";
export const DEFAULT_OLLAMA_BASE_URL = "http://localhost:11434";
export const DEFAULT_OPENAI_BASE_URL = "https://api.openai.com/v1";
export const DEFAULT_OPENROUTER_BASE_URL = "https://openrouter.ai/api/v1";
export const DEFAULT_MISTRAL_BASE_URL = "https://api.mistral.ai/v1";
export const VERTEX_CATEGORIZATION_BATCH_SIZE = 100;

const OPENAI_COMPATIBLE_DEFAULTS: Record<OpenAiCompatProvider, string> = {
  openai: DEFAULT_OPENAI_BASE_URL,
  openrouter: DEFAULT_OPENROUTER_BASE_URL,
  mistral: DEFAULT_MISTRAL_BASE_URL,
};

const OPENAI_COMPATIBLE_ENV_KEY: Record<OpenAiCompatProvider, string> = {
  openai: "OPENAI_API_KEY",
  openrouter: "OPENROUTER_API_KEY",
  mistral: "MISTRAL_API_KEY",
};

/** Raw commander opts after parse (before parseSensemakerModelOpts). */
export type SensemakerModelCliRawOpts = {
  adapter?: string;
  provider?: string;
  baseUrl?: string;
  apiKey?: string;
  modelName?: string;
  vertexProject?: string;
  vertexLocation?: string;
  keyFilename?: string;
  categorizationBatchSize?: string;
};

export interface SensemakerModelCliParsed {
  adapter: SensemakerAdapter;
  provider?: OpenAiCompatProvider;
  vertexProject?: string;
  vertexLocation: string;
  keyFilename?: string;
  modelName?: string;
  baseUrl: string;
  apiKey?: string;
  categorizationBatchSize?: number;
  categorizationBatchSizeCliProvided: boolean;
}

export function addSensemakerModelOptions(program: Command): Command {
  return program
    .option(
      "--adapter <adapter>",
      "LLM adapter: vertex (default), ollama, or openai-compatible.",
      "vertex"
    )
    .option(
      "--provider <provider>",
      "Provider preset for --adapter openai-compatible: openai, openrouter, or mistral."
    )
    .option(
      "--baseUrl <url>",
      "Base URL for the LLM HTTP API. Defaults depend on adapter/provider."
    )
    .option(
      "--apiKey <token>",
      "API key for openai-compatible backends. If omitted, provider-specific env var is used."
    )
    .option(
      "-m, --modelName <model>",
      `Model id for the selected adapter (Vertex default: ${DEFAULT_VERTEX_MODEL}; Ollama default: ${DEFAULT_OLLAMA_MODEL}; required for openai-compatible).`
    )
    .option(
      "-v, --vertexProject <project>",
      "Google Cloud project id (required when --adapter is vertex)."
    )
    .option(
      "--vertexLocation <location>",
      `Vertex location/region to use for --adapter vertex. Default: ${DEFAULT_VERTEX_LOCATION}.`,
      DEFAULT_VERTEX_LOCATION
    )
    .option(
      "-k, --keyFilename <file>",
      "Path to the service account key file for Vertex authentication."
    )
    .option(
      "--categorizationBatchSize <n>",
      "Statements per categorization batch. Used for Ollama only; Vertex always uses batch size 100."
    );
}

function normalizeAdapter(value: unknown): SensemakerAdapter {
  const s = String(value ?? "vertex").toLowerCase().trim();
  if (s === "vertex") {
    return "vertex";
  }
  if (s === "ollama") {
    return "ollama";
  }
  if (s === "openai-compatible") {
    return "openai-compatible";
  }
  throw new Error(
    `Invalid --adapter "${value}". Use "vertex", "ollama", or "openai-compatible".`
  );
}

function normalizeProvider(value: unknown): OpenAiCompatProvider | undefined {
  if (value === undefined || value === null || String(value).trim() === "") {
    return undefined;
  }
  const s = String(value).toLowerCase().trim();
  if (s === "openai" || s === "openrouter" || s === "mistral") {
    return s;
  }
  throw new Error(
    `Invalid --provider "${value}". Use "openai", "openrouter", or "mistral".`
  );
}

export function normalizeBaseUrl(url: string): string {
  return url.replace(/\/+$/, "");
}

export function parseSensemakerModelOpts(
  raw: SensemakerModelCliRawOpts,
  program: Command
): SensemakerModelCliParsed {
  const adapter = normalizeAdapter(raw.adapter);
  const provider = normalizeProvider(raw.provider);
  let categorizationBatchSize: number | undefined;
  if (
    raw.categorizationBatchSize !== undefined &&
    raw.categorizationBatchSize !== ""
  ) {
    const n = parseInt(String(raw.categorizationBatchSize), 10);
    if (Number.isNaN(n) || n < 1) {
      throw new Error("--categorizationBatchSize must be a positive integer.");
    }
    categorizationBatchSize = n;
  }

  const categorizationBatchSizeCliProvided =
    program.getOptionValueSource("categorizationBatchSize") === "cli";

  const baseUrl = getBaseUrl(adapter, provider, raw.baseUrl);

  return {
    adapter,
    provider,
    vertexProject: raw.vertexProject?.trim() || undefined,
    vertexLocation: raw.vertexLocation?.trim() || DEFAULT_VERTEX_LOCATION,
    keyFilename: raw.keyFilename,
    modelName: raw.modelName?.trim() || undefined,
    baseUrl,
    apiKey: raw.apiKey?.trim() || undefined,
    categorizationBatchSize,
    categorizationBatchSizeCliProvided,
  };
}

function getBaseUrl(
  adapter: SensemakerAdapter,
  provider: OpenAiCompatProvider | undefined,
  explicit?: string
): string {
  if (explicit && String(explicit).trim().length > 0) {
    return normalizeBaseUrl(String(explicit).trim());
  }
  if (adapter === "ollama") {
    return DEFAULT_OLLAMA_BASE_URL;
  }
  if (adapter === "openai-compatible" && provider) {
    return OPENAI_COMPATIBLE_DEFAULTS[provider];
  }
  return DEFAULT_OLLAMA_BASE_URL;
}

export function resolveApiKey(opts: SensemakerModelCliParsed): string | undefined {
  if (opts.apiKey) {
    return opts.apiKey;
  }
  if (opts.adapter !== "openai-compatible" || !opts.provider) {
    return undefined;
  }
  return process.env[OPENAI_COMPATIBLE_ENV_KEY[opts.provider]];
}

export function validateSensemakerModelOpts(opts: SensemakerModelCliParsed): void {
  if (opts.adapter === "vertex") {
    if (!opts.vertexProject) {
      throw new Error(
        "--vertexProject is required when --adapter is vertex. " +
          "For Ollama, use --adapter ollama."
      );
    }
  }
  if (opts.adapter === "ollama" || opts.adapter === "openai-compatible") {
    try {
      new URL(opts.baseUrl);
    } catch {
      throw new Error(
        `Invalid --baseUrl "${opts.baseUrl}". Expected a full URL (e.g. http://localhost:11434).`
      );
    }
  }
  if (opts.adapter === "openai-compatible") {
    if (!opts.provider) {
      throw new Error(
        '--provider is required when --adapter is openai-compatible. Use "openai", "openrouter", or "mistral".'
      );
    }
    if (!opts.modelName) {
      throw new Error(
        "--modelName is required when --adapter is openai-compatible."
      );
    }
    if (!resolveApiKey(opts)) {
      const envVar = OPENAI_COMPATIBLE_ENV_KEY[opts.provider];
      throw new Error(
        `API key is required for provider "${opts.provider}". Pass --apiKey or set ${envVar}.`
      );
    }
  }
}

export function warnCategorizationBatchSizeForVertex(
  opts: SensemakerModelCliParsed
): void {
  if (
    opts.adapter === "vertex" &&
    opts.categorizationBatchSizeCliProvided
  ) {
    console.warn(
      `[sensemaker] --categorizationBatchSize is ignored for Vertex; categorization uses fixed batch size ${VERTEX_CATEGORIZATION_BATCH_SIZE}.`
    );
  }
}

export function createModelFromCliOptions(opts: SensemakerModelCliParsed): Model {
  if (opts.adapter === "vertex") {
    const modelName = opts.modelName ?? DEFAULT_VERTEX_MODEL;
    return new VertexModel(
      opts.vertexProject!,
      opts.vertexLocation,
      modelName,
      opts.keyFilename
    );
  }
  const modelName = opts.modelName ?? DEFAULT_OLLAMA_MODEL;
  const batch =
    opts.categorizationBatchSize !== undefined
      ? opts.categorizationBatchSize
      : 5;
  if (opts.adapter === "ollama") {
    return new OllamaModel(opts.baseUrl, modelName, batch);
  }
  return new OpenAiCompatModel({
    baseUrl: opts.baseUrl,
    apiKey: resolveApiKey(opts)!,
    modelName: opts.modelName!,
    provider: opts.provider!,
  });
}

File: src/models/sensemaker_model_cli.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Shared CLI options for selecting SensemakingLlm backends (gemini / openai-compatible)."""

from __future__ import annotations

import argparse
import os
from dataclasses import dataclass, replace
from typing import Literal
from urllib.parse import urlparse

from src.models.sensemaking_llm import SensemakingLlm

SensemakerAdapter = Literal["gemini", "openai-compatible"]
OpenAiCompatProvider = Literal["openai", "openrouter", "mistral"]

DEFAULT_ADAPTER: SensemakerAdapter = "gemini"
DEFAULT_OPENAI_BASE_URL = "https://api.openai.com/v1"
DEFAULT_OPENROUTER_BASE_URL = "https://openrouter.ai/api/v1"
DEFAULT_MISTRAL_BASE_URL = "https://api.mistral.ai/v1"

OPENAI_COMPAT_DEFAULT_BASE_URL: dict[OpenAiCompatProvider, str] = {
    "openai": DEFAULT_OPENAI_BASE_URL,
    "openrouter": DEFAULT_OPENROUTER_BASE_URL,
    "mistral": DEFAULT_MISTRAL_BASE_URL,
}

OPENAI_COMPAT_ENV_KEYS: dict[OpenAiCompatProvider, str] = {
    "openai": "OPENAI_API_KEY",
    "openrouter": "OPENROUTER_API_KEY",
    "mistral": "MISTRAL_API_KEY",
}


@dataclass
class SensemakerModelConfig:
  """Parsed LLM adapter settings from CLI arguments."""

  adapter: SensemakerAdapter
  provider: OpenAiCompatProvider | None
  base_url: str
  model_name: str | None
  api_key: str | None
  openrouter_site_url: str | None
  openrouter_app_name: str | None


def normalize_base_url(url: str) -> str:
  """Strips trailing slashes from an API base URL."""
  return url.rstrip("/")


def normalize_adapter(value: str | None) -> SensemakerAdapter:
  """Validates and normalizes --adapter."""
  s = (value or DEFAULT_ADAPTER).lower().strip()
  if s == "gemini":
    return "gemini"
  if s == "openai-compatible":
    return "openai-compatible"
  raise ValueError(
      f'Invalid --adapter "{value}". Use "gemini" or "openai-compatible".'
  )


def normalize_provider(value: str | None) -> OpenAiCompatProvider | None:
  """Validates and normalizes --provider."""
  if value is None or str(value).strip() == "":
    return None
  s = str(value).lower().strip()
  if s in ("openai", "openrouter", "mistral"):
    return s  # type: ignore[return-value]
  raise ValueError(
      f'Invalid --provider "{value}". Use "openai", "openrouter", or "mistral".'
  )


def get_base_url(
    adapter: SensemakerAdapter,
    provider: OpenAiCompatProvider | None,
    explicit: str | None,
) -> str:
  """Resolves base URL from explicit flag or adapter/provider presets."""
  if explicit and str(explicit).strip():
    return normalize_base_url(str(explicit).strip())
  if adapter == "openai-compatible" and provider:
    return OPENAI_COMPAT_DEFAULT_BASE_URL[provider]
  return ""


def add_sensemaker_model_options(parser: argparse.ArgumentParser) -> None:
  """Registers shared LLM adapter flags on an ArgumentParser."""
  parser.add_argument(
      "--adapter",
      type=str,
      default=DEFAULT_ADAPTER,
      help=(
          'LLM adapter: "gemini" (default, Google AI Studio) or'
          ' "openai-compatible".'
      ),
  )
  parser.add_argument(
      "--provider",
      type=str,
      default=None,
      help=(
          'Provider preset when --adapter is openai-compatible: "openai",'
          ' "openrouter", or "mistral".'
      ),
  )
  parser.add_argument(
      "--base_url",
      type=str,
      default=None,
      help="API base URL override. Defaults depend on --adapter and --provider.",
  )
  parser.add_argument(
      "--api_key",
      type=str,
      default=None,
      help="API key override. Otherwise uses provider-specific env vars.",
  )
  parser.add_argument(
      "--openrouter_site_url",
      type=str,
      default=None,
      help="Optional HTTP-Referer header for OpenRouter attribution.",
  )
  parser.add_argument(
      "--openrouter_app_name",
      type=str,
      default=None,
      help="Optional X-Title header for OpenRouter attribution.",
  )


def parse_sensemaker_model_opts(args: argparse.Namespace) -> SensemakerModelConfig:
  """Builds SensemakerModelConfig from parsed argparse.Namespace."""
  adapter = normalize_adapter(getattr(args, "adapter", None))
  provider = normalize_provider(getattr(args, "provider", None))
  base_url = get_base_url(
      adapter, provider, getattr(args, "base_url", None)
  )
  api_key = getattr(args, "api_key", None)
  if api_key:
    api_key = api_key.strip() or None
  return SensemakerModelConfig(
      adapter=adapter,
      provider=provider,
      base_url=base_url,
      model_name=getattr(args, "model_name", None),
      api_key=api_key,
      openrouter_site_url=getattr(args, "openrouter_site_url", None),
      openrouter_app_name=getattr(args, "openrouter_app_name", None),
  )


def resolve_api_key(opts: SensemakerModelConfig) -> str | None:
  """Resolves API key from CLI or environment."""
  if opts.api_key:
    return opts.api_key
  if opts.adapter == "gemini":
    return os.getenv("GOOGLE_API_KEY")
  if opts.adapter == "openai-compatible" and opts.provider:
    return os.getenv(OPENAI_COMPAT_ENV_KEYS[opts.provider])
  return None


def validate_sensemaker_model_opts(opts: SensemakerModelConfig) -> None:
  """Raises ValueError if options are inconsistent or incomplete."""
  if opts.adapter == "openai-compatible":
    if not opts.provider:
      raise ValueError(
          '--provider is required when --adapter is openai-compatible. Use'
          ' "openai", "openrouter", or "mistral".'
      )
    if not opts.model_name:
      raise ValueError(
          "--model_name is required when --adapter is openai-compatible."
      )
    try:
      urlparse(opts.base_url)
      if not opts.base_url.startswith("http"):
        raise ValueError()
    except ValueError as e:
      raise ValueError(
          f'Invalid --base_url "{opts.base_url}". Expected a full URL.'
      ) from e
    if not resolve_api_key(opts):
      env_var = OPENAI_COMPAT_ENV_KEYS[opts.provider]
      raise ValueError(
          f'API key is required for provider "{opts.provider}". Pass'
          f" --api_key or set {env_var}."
      )
  elif opts.adapter == "gemini":
    # API key is resolved when constructing GenaiModel (GOOGLE_API_KEY or --api_key).
    pass


def resolve_cli_api_key_from_args(
    args: argparse.Namespace,
    explicit_api_key: str | None = None,
) -> str | None:
  """Resolves API key from explicit arg, ``--api_key``, or legacy ``--gemini_api_key``."""
  if explicit_api_key and str(explicit_api_key).strip():
    return str(explicit_api_key).strip()
  for attr in ("api_key", "gemini_api_key"):
    val = getattr(args, attr, None)
    if val and str(val).strip():
      return str(val).strip()
  return None


def create_llm_from_args(
    args: argparse.Namespace,
    *,
    model_name: str | None = None,
    api_key: str | None = None,
    max_llm_retries: int | None = None,
    stats_log_file: str | None = None,
) -> SensemakingLlm:
  """Creates a SensemakingLlm from argparse Namespace and shared adapter flags.

  Args:
    args: Parsed CLI arguments including adapter/provider flags.
    model_name: Overrides ``--model_name`` when the runner uses a different flag
      name (e.g. ``--eval_model_name``) or a per-stage model id.
    api_key: Explicit API key override; else uses ``--api_key``, legacy
      ``--gemini_api_key``, or provider env vars.
    max_llm_retries: Optional retry override for the backend.
    stats_log_file: Optional path for stage stats summaries.

  Returns:
    Configured GenaiModel or OpenAiCompatLlm instance.
  """
  opts = parse_sensemaker_model_opts(args)
  legacy_key = resolve_cli_api_key_from_args(args, api_key)
  if legacy_key:
    opts = replace(opts, api_key=legacy_key)
  effective_model = model_name or opts.model_name
  if getattr(args, "model_name", None) and not effective_model:
    effective_model = args.model_name
  if effective_model:
    opts = replace(opts, model_name=effective_model)
  validate_sensemaker_model_opts(opts)
  from src.models import llm_factory  # pylint: disable=import-outside-toplevel

  return llm_factory.create_sensemaking_llm(
      opts,
      max_llm_retries=max_llm_retries,
      stats_log_file=stats_log_file,
  )
